File: app/frontend/services/productService.ts
// API configuration
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

// Import session manager
import { triggerSessionExpired } from '../hooks/useSessionManager';

// Types matching backend schemas
export interface Product {
  id: string;
  name: string;
  description?: string;
  sku: string;
  barcode?: string;
  category: string;
  subcategory?: string;
  brand?: string;
  cost_price: number;
  selling_price: number;
  quantity_in_stock: number;
  min_stock_threshold: number;
  max_stock_threshold?: number;
  aisle?: string;
  shelf?: string;
  bin_location?: string;
  warehouse_id?: string;
  unit_of_measure: string;
  weight?: number;
  dimensions?: string;
  is_perishable: boolean;
  expiry_date?: string;
  days_until_expiry_warning: number;
  supplier_id?: string;
  supplier_sku?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  created_by?: string;
  stock_status: 'OUT_OF_STOCK' | 'LOW_STOCK' | 'NORMAL' | 'OVERSTOCK';
  is_low_stock: boolean;
  is_out_of_stock: boolean;
}

export interface ProductCreateData {
  name: string;
  description?: string;
  sku: string;
  barcode?: string;
  category: string;
  subcategory?: string;
  brand?: string;
  cost_price: number;
  selling_price: number;
  quantity_in_stock: number;
  min_stock_threshold: number;
  max_stock_threshold?: number;
  aisle?: string;
  shelf?: string;
  bin_location?: string;
  warehouse_id?: string;
  unit_of_measure: string;
  weight?: number;
  dimensions?: string;
  is_perishable: boolean;
  expiry_date?: string;
  days_until_expiry_warning: number;
  supplier_id?: string;
  supplier_sku?: string;
}

export interface ProductListResponse {
  products: Product[];
  total_count: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface ProductFilters {
  page?: number;
  per_page?: number;
  category?: string;
  subcategory?: string;
  brand?: string;
  supplier_id?: string;
  warehouse_id?: string;
  is_perishable?: boolean;
  stock_status?: string;
  aisle?: string;
  search?: string;
}

export interface DashboardMetrics {
  total_products: number;
  low_stock_count: number;
  out_of_stock_count: number;
  expiring_soon_count: number;
  total_value: number;
}

// Helper function to get auth token from localStorage
const getAuthToken = (): string | null => {
  if (typeof window !== 'undefined') {
    return localStorage.getItem('access_token');
  }
  return null;
};

// Helper function to make API requests
const apiRequest = async <T>(
  endpoint: string, 
  options: RequestInit = {}
): Promise<T> => {
  const url = `${API_BASE_URL}${endpoint}`;
  const token = getAuthToken();

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...(options.headers as Record<string, string> || {}),
  };

  // Only add Authorization header if token exists
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  const config: RequestInit = {
    headers,
    ...options,
  };

  try {
    const response = await fetch(url, config);
    
    if (!response.ok) {
      // Handle 401 Unauthorized - show session expired dialog
      if (response.status === 401) {
        console.warn('Authentication failed - showing session expired dialog');
        
        // Trigger the session expired dialog instead of throwing error
        triggerSessionExpired();
        throw new Error('Session expired');
      }
      
      const errorData = await response.json().catch(() => ({}));
      console.error('API Error Response:', errorData);
      
      // Handle different error response formats
      let errorMessage = `HTTP error! status: ${response.status}`;
      
      if (errorData.detail) {
        if (typeof errorData.detail === 'string') {
          errorMessage = errorData.detail;
        } else if (Array.isArray(errorData.detail)) {
          // Handle FastAPI validation errors
          const validationErrors = errorData.detail.map((err: { loc: string[]; msg: string }) => 
            `${err.loc.join('.')}: ${err.msg}`
          ).join(', ');
          errorMessage = `Validation errors: ${validationErrors}`;
        } else {
          errorMessage = JSON.stringify(errorData.detail);
        }
      } else if (errorData.message) {
        errorMessage = errorData.message;
      }
      
      throw new Error(errorMessage);
    }

    return await response.json();
  } catch (error) {
    console.error(`API request failed: ${endpoint}`, error);
    throw error;
  }
};

// Product Service Class
export class ProductService {
  // Get paginated list of products with filters
  static async getProducts(filters: ProductFilters = {}): Promise<ProductListResponse> {
    const searchParams = new URLSearchParams();
    
    Object.entries(filters).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        searchParams.append(key, value.toString());
      }
    });

    const endpoint = `/api/products?${searchParams.toString()}`;
    return apiRequest<ProductListResponse>(endpoint);
  }

  // Get single product by ID
  static async getProduct(id: string): Promise<Product> {
    return apiRequest<Product>(`/api/products/${id}`);
  }

  // Create new product
  static async createProduct(productData: ProductCreateData): Promise<Product> {
    return apiRequest<Product>('/api/products', {
      method: 'POST',
      body: JSON.stringify(productData),
    });
  }

  // Update existing product
  static async updateProduct(id: string, updates: Partial<Product>): Promise<Product> {
    return apiRequest<Product>(`/api/products/${id}`, {
      method: 'PUT',
      body: JSON.stringify(updates),
    });
  }

  // Delete product (soft delete)
  static async deleteProduct(id: string): Promise<{ message: string; success: boolean }> {
    return apiRequest<{ message: string; success: boolean }>(`/api/products/${id}`, {
      method: 'DELETE',
    });
  }

  // Adjust product stock
  static async adjustStock(
    id: string, 
    newQuantity: number, 
    reason: string, 
    notes?: string
  ): Promise<Product> {
    return apiRequest<Product>(`/api/products/${id}/stock`, {
      method: 'PATCH',
      body: JSON.stringify({
        product_id: id,
        new_quantity: newQuantity,
        reason,
        notes,
      }),
    });
  }

  // Get product by barcode
  static async getProductByBarcode(barcode: string): Promise<Product> {
    return apiRequest<Product>(`/api/products/barcode/${barcode}`);
  }

  // Get product by SKU
  static async getProductBySku(sku: string): Promise<Product> {
    return apiRequest<Product>(`/api/products/sku/${sku}`);
  }

  // Get all categories
  static async getCategories(): Promise<{ categories: Array<{ name: string; subcategories: string[] }> }> {
    return apiRequest<{ categories: Array<{ name: string; subcategories: string[] }> }>('/api/products/categories/list');
  }

  // Get dashboard metrics
  static async getDashboardMetrics(): Promise<DashboardMetrics> {
    return apiRequest<DashboardMetrics>('/api/dashboard/metrics');
  }

  // Get low stock items
  static async getLowStockItems(): Promise<{ items: Array<{ id: string; name: string; current_stock: number; min_threshold: number; location?: string }>; total_count: number }> {
    return apiRequest<{ items: Array<{ id: string; name: string; current_stock: number; min_threshold: number; location?: string }>; total_count: number }>('/api/dashboard/low-stock');
  }
}

// Error handling utility
export class ApiError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public details?: unknown
  ) {
    super(message);
    this.name = 'ApiError';
  }
}

// Success message helper
export const showSuccessMessage = (message: string) => {
  // For now, just console.log - replace with your toast/notification system
  console.log('✅ Success:', message);
};

// Error message helper
export const showErrorMessage = (error: unknown) => {
  const message = error instanceof Error ? error.message : 'An unexpected error occurred';
  // For now, just console.error - replace with your toast/notification system
  console.error('❌ Error:', message);
}; 

File: app/backend/services/product_service.py
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, func
from typing import List, Tuple, Optional
from datetime import datetime

from models.product import Product, InventoryTransaction
from models.schemas import (
    ProductCreate, 
    ProductUpdate, 
    ProductFilter,
    ProductResponse,
    TransactionType
)

class ProductService:
    def __init__(self, db: Session):
        self.db = db

    def create_product(self, product: ProductCreate, user_id: str) -> ProductResponse:
        """
        Create a new product in the database
        """
        db_product = Product(
            **product.dict(),
            created_by=user_id
        )
        
        self.db.add(db_product)
        self.db.commit()
        self.db.refresh(db_product)
        
        # Create initial stock transaction if quantity > 0
        if product.quantity_in_stock > 0:
            self._create_stock_transaction(
                product_id=db_product.id,
                transaction_type=TransactionType.IN,
                quantity=product.quantity_in_stock,
                reference_type="MANUAL",
                notes="Initial stock",
                user_id=user_id
            )
        
        return self._to_response(db_product)

    def get_product(self, product_id: str) -> Optional[ProductResponse]:
        """
        Get a product by ID
        """
        product = self.db.query(Product).filter(
            and_(Product.id == product_id, Product.is_active == True)
        ).first()
        
        return self._to_response(product) if product else None

    def get_product_by_sku(self, sku: str) -> Optional[ProductResponse]:
        """
        Get a product by SKU
        """
        product = self.db.query(Product).filter(
            and_(Product.sku == sku, Product.is_active == True)
        ).first()
        
        return self._to_response(product) if product else None

    def get_product_by_barcode(self, barcode: str) -> Optional[ProductResponse]:
        """
        Get a product by barcode
        """
        product = self.db.query(Product).filter(
            and_(Product.barcode == barcode, Product.is_active == True)
        ).first()
        
        return self._to_response(product) if product else None

    def get_products_paginated(
        self, 
        page: int = 1, 
        per_page: int = 20, 
        filters: Optional[ProductFilter] = None
    ) -> Tuple[List[ProductResponse], int]:
        """
        Get paginated list of products with filters
        """
        query = self.db.query(Product).filter(Product.is_active == True)
        
        # Apply filters
        if filters:
            if filters.category:
                query = query.filter(Product.category.ilike(f"%{filters.category}%"))
            
            if filters.subcategory:
                query = query.filter(Product.subcategory.ilike(f"%{filters.subcategory}%"))
            
            if filters.brand:
                query = query.filter(Product.brand.ilike(f"%{filters.brand}%"))
            
            if filters.supplier_id:
                query = query.filter(Product.supplier_id == filters.supplier_id)
            
            if filters.warehouse_id:
                query = query.filter(Product.warehouse_id == filters.warehouse_id)
            
            if filters.is_perishable is not None:
                query = query.filter(Product.is_perishable == filters.is_perishable)
            
            if filters.aisle:
                query = query.filter(Product.aisle.ilike(f"%{filters.aisle}%"))
            
            if filters.stock_status:
                if filters.stock_status == "OUT_OF_STOCK":
                    query = query.filter(Product.quantity_in_stock <= 0)
                elif filters.stock_status == "LOW_STOCK":
                    query = query.filter(
                        and_(
                            Product.quantity_in_stock > 0,
                            Product.quantity_in_stock <= Product.min_stock_threshold
                        )
                    )
                elif filters.stock_status == "OVERSTOCK":
                    query = query.filter(Product.quantity_in_stock >= Product.max_stock_threshold)
                elif filters.stock_status == "NORMAL":
                    query = query.filter(
                        and_(
                            Product.quantity_in_stock > Product.min_stock_threshold,
                            Product.quantity_in_stock < Product.max_stock_threshold
                        )
                    )
            
            if filters.search:
                search_term = f"%{filters.search}%"
                query = query.filter(
                    or_(
                        Product.name.ilike(search_term),
                        Product.description.ilike(search_term),
                        Product.barcode.ilike(search_term),
                        Product.sku.ilike(search_term)
                    )
                )
        
        # Get total count before pagination
        total_count = query.count()
        
        # Apply pagination
        offset = (page - 1) * per_page
        products = query.offset(offset).limit(per_page).all()
        
        return [self._to_response(product) for product in products], total_count

    def update_product(self, product_id: str, product_update: ProductUpdate) -> ProductResponse:
        """
        Update a product's information
        """
        product = self.db.query(Product).filter(
            and_(Product.id == product_id, Product.is_active == True)
        ).first()
        
        if not product:
            raise ValueError(f"Product with ID '{product_id}' not found")
        
        # Update only provided fields
        update_data = product_update.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(product, field, value)
        
        product.updated_at = datetime.utcnow()
        
        self.db.commit()
        self.db.refresh(product)
        
        return self._to_response(product)

    def delete_product(self, product_id: str) -> bool:
        """
        Soft delete a product (mark as inactive)
        """
        product = self.db.query(Product).filter(
            and_(Product.id == product_id, Product.is_active == True)
        ).first()
        
        if not product:
            return False
        
        product.is_active = False
        product.updated_at = datetime.utcnow()
        
        self.db.commit()
        return True

    def adjust_stock(
        self, 
        product_id: str, 
        new_quantity: int, 
        reason: str,
        user_id: str,
        notes: Optional[str] = None
    ) -> ProductResponse:
        """
        Adjust product stock quantity and create transaction record
        """
        product = self.db.query(Product).filter(
            and_(Product.id == product_id, Product.is_active == True)
        ).first()
        
        if not product:
            raise ValueError(f"Product with ID '{product_id}' not found")
        
        old_quantity = product.quantity_in_stock
        quantity_change = new_quantity - old_quantity
        
        # Update product quantity
        product.quantity_in_stock = new_quantity
        product.updated_at = datetime.utcnow()
        
        # Create inventory transaction
        transaction_type = TransactionType.IN if quantity_change > 0 else TransactionType.OUT
        
        self._create_stock_transaction(
            product_id=product_id,
            transaction_type=TransactionType.ADJUSTMENT,
            quantity=quantity_change,
            reference_type="MANUAL",
            notes=f"Stock adjustment: {reason}. {notes or ''}",
            user_id=user_id
        )
        
        self.db.commit()
        self.db.refresh(product)
        
        return self._to_response(product)

    def get_categories(self) -> List[dict]:
        """
        Get all unique categories and subcategories
        """
        categories_query = self.db.query(
            Product.category,
            Product.subcategory
        ).filter(Product.is_active == True).distinct().all()
        
        categories_dict = {}
        for category, subcategory in categories_query:
            if category not in categories_dict:
                categories_dict[category] = set()
            if subcategory:
                categories_dict[category].add(subcategory)
        
        # Convert to list format
        categories = []
        for category, subcategories in categories_dict.items():
            categories.append({
                "name": category,
                "subcategories": list(subcategories)
            })
        
        return categories

    def get_low_stock_products(self) -> List[ProductResponse]:
        """
        Get products with low stock levels
        """
        products = self.db.query(Product).filter(
            and_(
                Product.is_active == True,
                Product.quantity_in_stock <= Product.min_stock_threshold,
                Product.quantity_in_stock > 0
            )
        ).all()
        
        return [self._to_response(product) for product in products]

    def get_out_of_stock_products(self) -> List[ProductResponse]:
        """
        Get products that are out of stock
        """
        products = self.db.query(Product).filter(
            and_(
                Product.is_active == True,
                Product.quantity_in_stock <= 0
            )
        ).all()
        
        return [self._to_response(product) for product in products]

    def get_expiring_products(self, days: int = 7) -> List[ProductResponse]:
        """
        Get products expiring within specified days
        """
        from datetime import timedelta
        
        expiry_threshold = datetime.utcnow() + timedelta(days=days)
        
        products = self.db.query(Product).filter(
            and_(
                Product.is_active == True,
                Product.is_perishable == True,
                Product.expiry_date <= expiry_threshold,
                Product.expiry_date >= datetime.utcnow()
            )
        ).all()
        
        return [self._to_response(product) for product in products]

    def _create_stock_transaction(
        self,
        product_id: str,
        transaction_type: TransactionType,
        quantity: int,
        reference_type: str,
        notes: str,
        user_id: str,
        unit_cost: Optional[float] = None
    ):
        """
        Create an inventory transaction record
        """
        transaction = InventoryTransaction(
            product_id=product_id,
            transaction_type=transaction_type,
            quantity=quantity,
            unit_cost=unit_cost,
            reference_type=reference_type,
            notes=notes,
            created_by=user_id
        )
        
        self.db.add(transaction)

    def _to_response(self, product: Product) -> ProductResponse:
        """
        Convert Product model to ProductResponse schema
        """
        return ProductResponse(
            id=product.id,
            name=product.name,
            description=product.description,
            barcode=product.barcode,
            sku=product.sku,
            category=product.category,
            subcategory=product.subcategory,
            brand=product.brand,
            cost_price=product.cost_price,
            selling_price=product.selling_price,
            quantity_in_stock=product.quantity_in_stock,
            min_stock_threshold=product.min_stock_threshold,
            max_stock_threshold=product.max_stock_threshold,
            aisle=product.aisle,
            shelf=product.shelf,
            bin_location=product.bin_location,
            warehouse_id=product.warehouse_id,
            unit_of_measure=product.unit_of_measure,
            weight=product.weight,
            dimensions=product.dimensions,
            is_perishable=product.is_perishable,
            expiry_date=product.expiry_date,
            days_until_expiry_warning=product.days_until_expiry_warning,
            supplier_id=product.supplier_id,
            supplier_sku=product.supplier_sku,
            is_active=product.is_active,
            created_at=product.created_at,
            updated_at=product.updated_at,
            created_by=product.created_by,
            stock_status=product.stock_status,
            is_low_stock=product.is_low_stock,
            is_out_of_stock=product.is_out_of_stock
        ) 